### PR TITLE
0.3.1 RC1

### PR DESCRIPTION
**Pull request by jlmeunier**
I'd like to push this version, which is compatible with python 2 and 3, all tests are passing, it includes the CRF model I contributed, and it works with the latest AD3.

Version number: pystruct was in version 0.2.5, with a mention of 0.3 in the CHANGELOG. So I'm proposing 0.3.1

### CODE REVIEW

**Comment by amueller** [comment]
lol sounds like I was being very consistent. You can also do 0.4 but either is good.

**Comment by amueller** [comment]
maybe squash and merge for a nicer history?

File: examples/plot_hidden_snakes.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+    JL Meunier - January 2017
+    
+    Developed for the EU project READ. The READ project has received funding 
+    from the European Union's Horizon 2020 research and innovation programme 
+    under grant agreement No 674943
+    
+    Copyright Xerox
+
+This example uses the snake dataset introduced in
```
Comment (amueller) [inline examples/plot_hidden_snakes.py]: wow I think they didn't expect this snake to live on so long lol.

File: pystruct/inference/inference_methods.py [hunk 2]
```diff
@@ -350,30 +378,175 @@ def inference_ad3(unary_potentials, pairwise_potentials, edges, relaxed=False,
...
+        The READ project has received funding from the European Union's Horizon
+        2020 research and innovation programme under grant agreement No 674943.
+    
+    Returns
+    -------
+    labels : nd-array
+        Approximate (usually) MAP variable assignment.
+        If relaxed=False, this is a tuple of unary and edge 'marginals'.
+
+    Code written on Feb 2017 to deal with multiple node types, by JL Meunier,
```
Comment (amueller) [inline pystruct/inference/inference_methods.py]: I think it's more maintainable to have these either at the top of the file in the copyright header or in the readme (where it already is). If everybody that added/changed a method would add their affiliation to the docstring the scikit-learn docs would be unreadable now.

Comment (amueller) [reply pystruct/inference/inference_methods.py]: (I don't have a very strong opinion though).

File: pystruct/tests/pytest-0.3.3.log [hunk 3]
```diff
@@ -0,0 +1,703 @@
+============================= test session starts ==============================
```
Comment (amueller) [inline pystruct/tests/pytest-0.3.3.log]: did you mean to commit this?

File: pystruct/tests/test_learners/test_crammer_singer_svm.py [hunk 4]
```diff
@@ -188,14 +188,13 @@ def test_class_weights():
...
-    raise SkipTest("class weight test needs update")
     from sklearn.svm import LinearSVC
     X, Y = make_blobs(n_samples=210, centers=3, random_state=1, cluster_std=3,
                       shuffle=False)
     X = np.hstack([X, np.ones((X.shape[0], 1))])
-    X, Y = X[:170], Y[:170]
 
-    weights = len(Y) / (np.bincount(Y) * len(np.unique(Y)))
+    weights = 1. / np.bincount(Y)
+    weights *= len(weights) / np.sum(weights)
```
Comment (amueller) [inline pystruct/tests/test_learners/test_crammer_singer_svm.py]: I think I fixed this in master since then. The formula changed.

File: setup.py [hunk 5]
```diff
@@ -42,6 +42,6 @@
                    'Programming Language :: Python :: 2.6',
                    'Programming Language :: Python :: 2.7',
                    'Programming Language :: Python :: 3',
-                   'Programming Language :: Python :: 3.3',
+                   'Programming Language :: Python :: 3.6',
```
Comment (amueller) [inline setup.py]: I think we can remove 2.6 lol.

File: pystruct/inference/inference_methods.py (hunk 2 above)
Comment (jlmeunier) [reply pystruct/inference/inference_methods.py]: agree => removed

File: pystruct/tests/pytest-0.3.3.log (hunk 3 above)
Comment (jlmeunier) [reply pystruct/tests/pytest-0.3.3.log]: at some point yes, Now, no => removed

File: examples/plot_hidden_snakes.py (hunk 1 above)
Comment (jlmeunier) [reply examples/plot_hidden_snakes.py]: :))

File: pystruct/tests/test_learners/test_crammer_singer_svm.py (hunk 4 above)
Comment (jlmeunier) [reply pystruct/tests/test_learners/test_crammer_singer_svm.py]: I'm not clear on those weights. 
Which formula do you refer to?
BTW 1: np.bincount return an array with a 0 => division by zero
BTW 2: I never managed to use weights in pystruct to cope with unbalanced labels. Uniform weighting generally is best. :-O

**Comment by jlmeunier** [comment]
Squashing is tricky... 
Tried to squash in my fork, but got conflicts (!) (on setup.py)
I do not dare to squash in this repo...

I could do a new fork from pystruct master, prepare 0.3 there, do 1 commit and make a new pull request.
Not so elegant but it will work for sure...
Any other idea or info??
Thanks

**Comment by jlmeunier** [comment]
Squashing leads me in dealing with several conflicts. So no squashing. But a new PR from another repo (for the exact same code)
